File: src/main/java/gargoyle/ct/log/Logger.java
package gargoyle.ct.log;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

public interface Logger {

    String ROOT_LOGGER_NAME = "";

    void log(@NotNull Level level, @Nullable Throwable exception, @Nullable String pattern, Object... arguments);

    boolean isLoggable(@NotNull Level level);

    default boolean isTraceEnabled() {
        return isLoggable(Level.TRACE);
    }

    default boolean isDebugEnabled() {
        return isLoggable(Level.DEBUG);
    }

    default boolean isInfoEnabled() {
        return isLoggable(Level.INFO);
    }

    default boolean isWarningEnabled() {
        return isLoggable(Level.WARNING);
    }

    default boolean isErrorEnabled() {
        return isLoggable(Level.ERROR);
    }

    default void trace(Throwable exception, String pattern, Object... arguments) {
        log(Level.TRACE, exception, pattern, arguments);
    }

    default void trace(String pattern, Object... arguments) {
        log(Level.TRACE, null, pattern, arguments);
    }

    default void debug(Throwable exception, String pattern, Object... arguments) {
        log(Level.DEBUG, exception, pattern, arguments);
    }

    default void debug(String pattern, Object... arguments) {
        log(Level.DEBUG, null, pattern, arguments);
    }

    default void error(Throwable exception, String pattern, Object... arguments) {
        log(Level.ERROR, exception, pattern, arguments);
    }

    default void error(String pattern, Object... arguments) {
        log(Level.ERROR, null, pattern, arguments);
    }

    default void info(Throwable exception, String pattern, Object... arguments) {
        log(Level.INFO, exception, pattern, arguments);
    }

    default void info(String pattern, Object... arguments) {
        log(Level.INFO, null, pattern, arguments);
    }

    default void log(@NotNull Level level, String pattern, Object... arguments) {
        log(level, null, pattern, arguments);
    }

    default void warn(Throwable exception, String pattern, Object... arguments) {
        log(Level.WARNING, exception, pattern, arguments);
    }

    default void warn(String pattern, Object... arguments) {
        log(Level.WARNING, null, pattern, arguments);
    }
}

File: src/main/kotlin/gargoyle/ct/log/Logger.kt
package gargoyle.ct.log

interface Logger {
    fun log(level: Level, exception: Throwable?, pattern: String, vararg arguments: Any)
    fun isLoggable(level: Level): Boolean
    val isTraceEnabled: Boolean
        get() = isLoggable(Level.TRACE)
    val isDebugEnabled: Boolean
        get() = isLoggable(Level.DEBUG)
    val isInfoEnabled: Boolean
        get() = isLoggable(Level.INFO)
    val isWarningEnabled: Boolean
        get() = isLoggable(Level.WARNING)
    val isErrorEnabled: Boolean
        get() = isLoggable(Level.ERROR)

    fun trace(exception: Throwable?, pattern: String, vararg arguments: Any) {
        log(Level.TRACE, exception, pattern, *arguments)
    }

    fun trace(pattern: String, vararg arguments: Any) {
        log(Level.TRACE, null, pattern, *arguments)
    }

    fun debug(exception: Throwable?, pattern: String, vararg arguments: Any) {
        log(Level.DEBUG, exception, pattern, *arguments)
    }

    fun debug(pattern: String, vararg arguments: Any) {
        log(Level.DEBUG, null, pattern, *arguments)
    }

    fun error(exception: Throwable?, pattern: String, vararg arguments: Any) {
        log(Level.ERROR, exception, pattern, *arguments)
    }

    fun error(pattern: String, vararg arguments: Any) {
        log(Level.ERROR, null, pattern, *arguments)
    }

    fun info(exception: Throwable?, pattern: String, vararg arguments: Any) {
        log(Level.INFO, exception, pattern, *arguments)
    }

    fun info(pattern: String, vararg arguments: Any) {
        log(Level.INFO, null, pattern, *arguments)
    }

    fun log(level: Level, pattern: String, vararg arguments: Any) {
        log(level, null, pattern, *arguments)
    }

    fun warn(exception: Throwable?, pattern: String, vararg arguments: Any) {
        log(Level.WARNING, exception, pattern, *arguments)
    }

    fun warn(pattern: String, vararg arguments: Any) {
        log(Level.WARNING, null, pattern, *arguments)
    }

    companion object {
        const val ROOT_LOGGER_NAME = ""
    }
}
